pybossa/view/projects.py

```diff
@@ -1698,6 +1699,110 @@ def bulk_priority_update(short_name):
     except Exception as e:
         return ErrorStatus().format_exception(e, 'priorityupdate', 'POST')
 
+@blueprint.route('/<short_name>/tasks/assign-workersupdate', methods=['POST'])
+def bulk_update_assign_worker(short_name):
+    import copy
+    from sqlalchemy.orm.attributes import flag_modified
+
+
+    response = {}
+    project, owner, ps = project_by_shortname(short_name)
+    data = json.loads(request.data)
+
+    if data.get("add") is None and data.get("remove") is None:
+        # read data and return users
+
+        task_id = data.get("taskId")
+        bulk_update = False
+        assign_user_emails = []
+        if task_id:
+            # use filters tp populate user list
+            t = task_repo.get_task_by(project_id=project.id,
+                                        id=int(task_id))
+            assign_users = []
+            if t.user_pref:
+                assign_user_emails = set(t.user_pref.get("assign_user", []))
+                for user_email in assign_user_emails:
+                    fullname = user_repo.search_by_email(user_email).fullname
+                    assign_users.append({'fullname': fullname, 'email': user_email})
+            response['assign_users'] = assign_users
+        else:
+            bulk_update = True
+            args = parse_tasks_browse_args(json.loads(data.get('filters', '')))
+            tasks = task_repo.get_tasks_by_filters(project, args)
+            task_ids = [t.id for t in tasks]
+            assign_user_emails = set()
+
+            for task_id in task_ids:
+                t = task_repo.get_task_by(project_id=project.id,
+                                        id=int(task_id))
+                assign_user_emails = assign_user_emails.union(set(t.user_pref.get("assign_user", [])))
+            assign_users = []
+            for user_email in assign_user_emails:
+                user = user_repo.search_by_email(user_email)
+                if not user:
+                    fullname = user_email + ' (user not found)'
+                else:
+                    fullname = user.fullname 
+                assign_users.append({'fullname': fullname, 'email': user_email})
```

Comment:
Looks like there's scope of consolidating the common code here; specifically turning user email ids from assign_user list to fullname, email list. Hence, code can be refactored something on similar lines as mentioned below
```python
assign_user_emails, assign_users = [], []
if task_id:
    assign_user_emails = set(t.user_pref.get("assign_user", []))
else:
       ... handle bulk task obtaining tasks from filter
        for task_id in task_ids:
                t = task_repo.get_task_by(project_id=project.id, id=int(task_id))
                assign_user_emails = assign_user_emails.union(set(t.user_pref.get("assign_user", [])))
for user_email in assign_user_emails:
     fullname = user_repo.search_by_email(user_email).fullname
     assign_users.append({'fullname': fullname, 'email': user_email})
response['assign_users'] = assign_users
```

Same code refactoring may be applicable under `POST` request section as well.

Reply:
great idea